core/src/main/scala/org/apache/spark/scheduler/TaskAssigner.scala

```diff
@@ -0,0 +1,233 @@
+ * information. The contract between [[org.apache.spark.scheduler.TaskScheduler TaskScheduler]]
+ * and TaskAssigner is as follows. First, TaskScheduler invokes construct() of TaskAssigner to
+ * initialize the its internal worker states at the beginning of resource offering. Before each
+ * round of task assignment for a taskset, TaskScheduler invoke the init() of TaskAssigner to
+ * initialize the data structure for the round. When performing real task assignment,
+ * hasNext()/getNext() is used by TaskScheduler to check the worker availability and retrieve
+ * current offering from TaskAssigner. Then offerAccepted is used by TaskScheduler to notify
+ * the TaskAssigner so that TaskAssigner can decide whether the current offer is valid or not for
+ * the next request. After task assignment is done, TaskScheduler invokes the tasks() to
+ * retrieve all the task assignment information, and eventually, invokes reset() method so that
+ * TaskAssigner can cleanup its internal maintained resources.
+ */
+
+private[scheduler] abstract class TaskAssigner {
+  var offer: Seq[OfferState] = _
+  var CPUS_PER_TASK = 1
+
+  def withCpuPerTask(CPUS_PER_TASK: Int): Unit = {
+    this.CPUS_PER_TASK = CPUS_PER_TASK
+  }
+
+  // The final assigned offer returned to TaskScheduler.
+  final def tasks: Seq[ArrayBuffer[TaskDescription]] = offer.map(_.tasks)
+
+  // Invoked at the beginning of resource offering to construct the offer with the workoffers.
+  def construct(workOffer: Seq[WorkerOffer]): Unit = {
+    offer = workOffer.map(o => new OfferState(o))
+  }
+
+  // Invoked at each round of Taskset assignment to initialize the internal structure.
+  def init(): Unit
+
+  // Whether there is offer available to be used inside of one round of Taskset assignment.
+  def hasNext: Boolean
+
+  // Returned the next assigned offer based on the task assignment strategy.
+  def getNext(): OfferState
+
+  // Invoked by the TaskScheduler to indicate whether the current offer is accepted or not so that
+  // the assigner can decide whether the current worker is valid for the next offering.
+  def offerAccepted(assigned: Boolean): Unit
+
+  // Invoked at the end of resource offering to release internally maintained resources.
+  // Subclass is responsible to release its own private resources.
+  def reset(): Unit = {
+    offer = null
+  }
+}
+
+object TaskAssigner extends Logging {
+  private val roundrobin = classOf[RoundRobinAssigner].getCanonicalName
+  private val packed = classOf[PackedAssigner].getCanonicalName
+  private val balanced = classOf[BalancedAssigner].getCanonicalName
+  private val assignerMap: Map[String, String] =
+    Map("roundrobin" -> roundrobin,
+      "packed" -> packed,
+      "balanced" -> balanced)
+
+  def init(conf: SparkConf): TaskAssigner = {
+    val assignerName = conf.get(config.SPARK_SCHEDULER_TASK_ASSIGNER.key, "roundrobin")
+      .toLowerCase()
+    val className = assignerMap.getOrElse(assignerName, roundrobin)
```

Comment:
After discussing, it is better to keep this restriction. Please refer to last PR
https://github.com/apache/spark/pull/15218#issuecomment-249237597